### PR TITLE
Add More Branch Filtering

### PR DESCRIPTION
**Pull request by RPGillespie6**
Description:
- Add new option `--include-br-lines-starting-with` for white listing branch coverage statements
- For example, `--include-br-lines-starting-with if else` would only include branch coverage for statements starting with `if` or `else`
- Standardized logging to always use `.format()`

### CODE REVIEW

**Comment by blueyed** [comment]
This appears to require you to use the `--include-br-lines-starting-with` in the first place to get branch coverage?!
I am using `cd build; fastcov -b --lcov -o report.info` - and without `--include-br-lines-starting-with if else` at the end I only see 6 branches (in some autogenerated, not sure why those show up then - worth investigating probably).
With the option added there are "32844/61088" branches.

File: fastcov.py [hunk 1]
```diff
@@ -196,9 +194,11 @@ def exclMarkerWorker(fastcov_sources, chunk, exclude_branches_sw):
         end_line = 0
         with open(source) as f:
             for i, line in enumerate(f, 1): #Start enumeration at line 1
-                # Exclude branches starting with...
-                for e in exclude_branches_sw:
-                    if line.lstrip().startswith(e) and str(i) in fastcov_sources[source]["branches"]:
+                if str(i) in fastcov_sources[source]["branches"]:
+                    if all(not line.lstrip().startswith(e) for e in include_branches_sw): # Include branches starting with...
```
Comment (blueyed) [inline fastcov.py]: This should maybe only be done if there is a whitelist in the first place?

**Comment by RPGillespie6** [comment]
By default, `-b` tries to filter "noisy" branches generated by GCC. In your case you would want to do:

`cd build; fastcov -B --lcov -o report.info - and without --include-br-lines-starting-with if else` (note capital `B`)

**Comment by RPGillespie6** [comment]
If you just want pure, unaltered branch coverage, just do:

`fastcov -B --lcov -o report.info`

**Comment by blueyed** [comment]
`-b` is fine I guess, but the problem looks like "whitelist is empty, and therefore `all([])` is true always, no?

The 6 branches I am still seeing then are btw:
```
(Pdb++) pp fastcov_sources[source]["branches"]
{'485': [0, 0], '489': [0, 0, 0, 0]}
(Pdb++) source
'/home/user/Vcs/neovim/build/src/nvim/auto/funcs.generated.h.gperf'
```
(I would assume those to be filtered then also - so there is likely another issue there)

**Comment by blueyed** [comment]
The 2nd issue is due to the file only having 484 lines.  Initially it is `{'481': [0, 0, 0, 0], '485': [0, 0], '489': [0, 0, 0, 0]} `.

File: fastcov.py (hunk 1 above)
Comment (RPGillespie6) [reply fastcov.py]: Oh, I see what you are saying now -

Should be `if include_branches_sw and all(...)`

This is quite possibly a bug, let me run a few tests.

Comment (RPGillespie6) [reply fastcov.py]: You're right, I need to setup Travis CI to prevent regressions in the future!

Comment (blueyed) [reply fastcov.py]: Ok, cool - I wasn't sure if I only misunderstood the new feature in the beginning.. ;)

I also suggest using pytest then.. :)
How do you run the existing test?

Comment (RPGillespie6) [reply fastcov.py]: Right now I run existing unit tests (of which there are not many) manually. I also manually run fastcov against nlohmann/json to make sure coverage hasn't changed.

I plan on adding black box testing that actually runs fastcov against CMake projects with predetermined coverage and branching.

Please check out #16 to see if it resolves your issues (I will test a bit more before merging).

Comment (blueyed) [reply fastcov.py]: Yes, works.

**Comment by blueyed** [comment]
Should I/we file a new issue for the last comment, i.e. I think it should ignore branch coverage for non-existing lines - not sure what the problem there is though really.

**Comment by RPGillespie6** [comment]
I'm wondering if it is allocation-related branching generated by GCC at compile time or something? You can raise a issue just for tracking, we should definitely exclude impossible-to-reach branches in `-b` (not with `-B` though)

**Pull request merged**